lib/clients/ZookeeperManager.js

```diff
@@ -0,0 +1,323 @@
+const zookeeper = require('node-zookeeper-client');
+const EventEmitter = require('events');
+const async = require('async');
+
+class ZookeeperManager extends EventEmitter {
+    /**
+     * Constructs an instance of the ZookeeperManager.
+     *
+     * @param {string} connectionString - The connection string for the ZooKeeper server.
+     * @param {Object} options - Configuration options for the ZooKeeper client.
+     * @param {Logger} log - An instance of a logging utility.
+     */
+    constructor(connectionString, options, log) {
+        super();
+
+        this.connectionString = connectionString;
+        this.log = log;
+        this.options = options;
+        this.client = null;
+
+        this._connect();
+    }
+
+    /**
+     * Establishes a connection to the ZooKeeper server.
+     *
+     * This method initializes a new ZooKeeper client using the provided connection string
+     * and options. It sets up event listeners for various client states such as 'connected',
+     * 'disconnected', and 'expired'. The method handles reconnection logic in case of
+     * client expiration and forwards ZooKeeper events through this class's EventEmitter interface.
+     *
+     * If 'autoCreateNamespace' is set in the options and the base path does not exist in ZooKeeper,
+     * this method tries to create the necessary path (namespace) in ZooKeeper.
+     *
+     * @method _connect
+     * @return {undefined}
+     */
+    _connect() {
+        // clean up exists client before reconnect
+        if (this.client) {
+            // close and clean up the existing ZooKeeper client connection.
+            this.close();
+            this.client.removeAllListeners();
+        }
+
+        this.client = zookeeper.createClient(this.connectionString, this.options);
+
+        let nsIndex;
+        let namespace;
+        let rootZkClient;
+        async.series([
+            // Check zookeeper client is connected
+            next => this.client.once('connected', next),
+            // Create namespace if it exists and options.autoCreateNamespace is true
+            next => {
+                // TODO: ARTESCA-10337 The 'autoCreateNamespace' functionality is currently specific to
+                // Artesca and may be removed in future versions once the Zenko Operator can handle base path creation.
+                // Once removed, we can simply rely on the 'connected' state instead of the 'ready' state and
+                // stop listening on 'error' event.
+                if (this.options && this.options.autoCreateNamespace) {
+                    nsIndex = this.connectionString.indexOf('/');
+                    namespace = this.connectionString.slice(nsIndex);
+                    if (nsIndex > -1 && namespace !== '/') {
```

Comment:
The connection string could look like `1.2.3.4:2181`, I don't think we would even see a `/` in this case.

Reply:
And this as well: https://github.com/scality/backbeat/pull/2471/files#diff-5d37e364a4ba55f3567e02ae1eaca3e7304140cf610d9c5da922751d5fc309f9R79

Reply:
Ah yes I missed the `nsIndex > -1` check, we precompute the `namespace` though without checking if there's a namespace, but in this case calling `slice(-1)` should not cause issues except it will return "garbage" in `namespace` but we don't use it. Might be slightly cleaner to compute `namespace` only if `nsIndex > -1` but it's minor and not blocking for me.